### DISCUSSION TITLE
Using a live-reload instead of browsersync for minimal projects

### DISCUSSION THREAD

**Discussion by mdwagner** [luckyframework/lucky#1635] [Feature requests & Ideas]
### Goal
For minimal-type Lucky projects, having the option to _not_ use browsersync would be great, as it could allow for a lower barrier to entry when starting a Lucky project.

I have already looked into this a bit, basically creating a [POC on what a live-reload without browsersync](https://github.com/mdwagner/importmap-example-lucky/blob/master/tasks/live_reload.cr) could look like, along with limitations it would have.

### How it works
Instead of starting a new Process with the browsersync executable, I start an HTTP::Server running a WebSocket server (although this could be an EventStream server alternatively). Then whenever it would need to refresh the browser, it sends a message to any WebSocket connections. These WebSocket connections would just call `location.reload()` whenever they get a message. [Here's an example of how I'm doing in a Lucky project](https://github.com/mdwagner/importmap-example-lucky/blob/master/src/components/shared/layout_head.cr#L37).

### Limitations
It actually works pretty well, but I have found that if you don't fingerprint your assets, the JS files don't change. This _can_ be solved with more work than what I've done here, but for just Crystal code changes it works great.

Combing this with something like [Alpine.js](https://alpinejs.dev/), where you don't typically do much JS development, it works just fine. But if you were trying using [React](https://reactjs.org/), this wouldn't work that well in development. This actually makes sense with the Goal, as that wouldn't be a minimal-type of project, so you might already be reaching for more complex tools anyway.

Let me know if you would like to explore this option more. I'm also on the Discord server.

**Comment by matthewmcgarvey** [comment]
Dude! Nice! Love the way your socket server is used to just capture the websockets! For some reason it's a bit mind bending that it's a whole little application that's just inlined in that class 🤯

**Comment by jwoertink** [comment]
I think this is definitely the way to go along with what @matthewmcgarvey did with Nox. There's always going to be someone that's like "I don't like that", or "I'd rather use X", but with Lucky providing something very basic and simple out of the box with the (easy-ish) option to "escape hatch" out to whatever, then I think it's a perfect solution.

The other thing too I just thought about, maybe we can use that server to connect on a socket as opposed to some port? That would help deal with the issue I always run in to where it says "something is already running on port 3001". If we had a way to not even worry about the ports, then it would be smooth. 

One thought I had as far as implementation/configuration was https://github.com/luckyframework/lucky/issues/920 where we could just use our `config/watch.yml`. That file is only used in development, and doesn't really do a whole lot right now. New generated apps could add something like

```yml
host: 127.0.0.1
port: 5000
watch_files:
  - src/js/*.js
  - src/css/*.scss
  - src/**/*.cr
  - config/*.cr
```

No clue if that's a good idea or not, but it's an idea...

Related: https://github.com/luckyframework/lucky/issues/1255
Related: https://github.com/luckyframework/lucky/issues/857

**Reply by mdwagner** [reply]
Thats a pretty good idea using the `config/watch.yml` file! I'll have to implement that soon!

I'm not so sure about using a socket though, as a quick google search shows its either pretty complicated or not supported, unless you can point to a good example of it. However, we could allow multiple ports in case one is taken. This would mean the JS might try opening a few different connections (because it wouldnt know which one is actually accepting connections), but it would be a decent workaround without much effort.

**Reply by jwoertink** [reply]
Maybe we try the first port (e.g. 3000), and if we see it's already in use, we just increment it until it works?

**Reply by mdwagner** [reply]
Finding a port is not really the problem, it's making sure the browser knows which port to try connecting to.

What I'm thinking we could do is check the port before we build the application the first time (until it finds one available), then we could store it in the watch.yml file and have the application look at that file for what port to use. 

So, we could specify what ports are acceptable (ports property), and then a port that is the current one used (current_port property).

Or we could just store the current port used in a tmp file locally in the project directory. The application would just look at that file for what port to use.

**Reply by jwoertink** [reply]
Maybe we don't have to scan at all. The current implementation doesn't really allow you to change that browsersync port. If we set that port as a value in the watch file, and we see that port is taken during boot, then we can just show a helpful error that says something like

> Could not boot Whatever this is called on port XXXX. That port seems to be in use already.
>
> Try this...
>   open config/watch.yml and change the whatever_port to YYYY

Then it's not really on us to make all that happen. We tried, it failed, here's how you fix it... What do you think about that?

**Reply by mdwagner** [reply]
I like that. Much easier to implement and is pretty much what I would do anyway in that situation.

**Reply by mdwagner** [reply]
@jwoertink https://github.com/mdwagner/importmap-example-lucky/commit/2fd5baa6e75aa5d652883bae85b11af516faa37c

**Comment by mdwagner** [comment]
So, besides putting up a PR for this feature, it does raise a few questions:
- should this feature be a different LuckyTask than Watch?
- if not, should the default option of Watch be this with an opt-out to browersync?

I don't necessarily think the browsersync implementation should be removed if people want it. I'm pretty confident both can be supported.

Although, if we went the other way where we add a new LuckyTask, we could just say it's experimental.

**Reply by jwoertink** [reply]
My thought was Lucky provides something basic that "gets the job done"... For people that want more control (i.e. using browsersync), there would be an escape hatch. It wouldn't really be like "which one of these options do  you want?", it would be more like "use this option.... if you don't like it, here's the off button so you can do your own thing". Does that make sense?

We just want to make sure that when someone does want to use another option, that it's not super difficult with all kinds of monkey patches.

**Reply by mdwagner** [reply]
do you think Watch should be extensible?

**Comment by mdwagner** [comment]
Hey everyone, just wanted to say I plan to get back to working on this soon.
